astrokat/observe_main.py

```diff
@@ -220,14 +220,34 @@ def cadence_target(target_list):
 
 def above_horizon(katpt_target, horizon=20.0, duration=0.0):
     """Check target visibility."""
-    [azim, elev] = katpt_target.azel(time.time())
+    # use local copies so you do not overwrite target time attribute
+    if 'azel' in katpt_target.tags:
+        return True
+
+    # must be celestial target (ra, dec)
+    local_target = katpt_target.body.copy()
+    local_observer = katpt_target.antenna.observer.copy()
```

Comment:
In katpoint's [Target.azel](https://github.com/ska-sa/katpoint/blob/e6a9efdff0c2b4d87e33e0a4a8a1f81117ea7117/katpoint/target.py#L310) method, we can pass an antenna object.  If that object is a copy, then it will modify that copy instead of the original.

So something like this:
```
local_antenna = copy.deepcopy(katpt_target.antenna)
azim, elev = katpt_target.azel(time.time(), antenna=local_antenna)
...
azim, elev = katpt_target.azel(time.time() + duration, antenna=local_antenna)
```

Reply:
General clean up of the `above_horizon` function implementation.
Target and observer are now passed explicitly with the observer being a copy of the antenna object.
This seems to work and keeps the `katpt_target` contained in the `run_observation` and explicitly requires ephem objects in the horizon function.